openrtb_ext/convert_up.go

```diff
@@ -25,6 +25,11 @@ func ConvertUpTo26(r *RequestWrapper) error {
 	// eid
 	moveEIDFrom25To26(r)
 
+	// imp
+	for _, imp := range r.GetImp() {
+		moveRewardedFromPrebidExtTo26(imp)
+	}
+
```

Comment:
I realize this isn't part of the current pull request but, are there plans to use `convertUpEnsureExt` somewhere else? If not, should `moveSupplyChainFrom24To25(r)`, `moveSupplyChainFrom25To26(r)`, `moveGDPRFrom25To26(r)`, `moveConsentFrom25To26(r)`, `moveUSPrivacyFrom25To26(r)`, `moveEIDFrom25To26(r)` and `moveRewardedFromPrebidExtTo26(imp)` return errors in order to optimize in:
 - Only traversing the `[]Imp` array once
 - Maintain one less function
 - Minimize the number of calls to `r.GetRequestExt()`, `r.GetSourceExt()`, `r.GetRegExt()` and `r.GetUserExt()`

```
 9   func ConvertUpTo26(r *RequestWrapper) error {
10 -     if err := convertUpEnsureExt(r); err != nil {
11 -         return err
12 -     }
13   
14       // schain
15 -     moveSupplyChainFrom24To25(r)
16 -     moveSupplyChainFrom25To26(r)
   +     if err := moveSupplyChainFrom24To25(r); err != nil {
   +         return err
   +     }
   +     if err := moveSupplyChainFrom25To26(r); err != nil {
   +         return err
   +     }
17   
18       // gdpr
19 -     moveGDPRFrom25To26(r)
20 -     moveConsentFrom25To26(r)
   +     if err := moveGDPRFrom25To26(r); err != nil {
   +         return err
   +     }
   +     if err := moveConsentFrom25To26(r); err != nil {
   +         return err
   +     }
21   
22       // ccpa
23 -     moveUSPrivacyFrom25To26(r)
   +     if err := moveUSPrivacyFrom25To26(r); err != nil {
   +         return err
   +     }
24   
25       // eid
26 -     moveEIDFrom25To26(r)
   +     if err := moveEIDFrom25To26(r); err != nil {
   +         return err
   +     }
27   
28       // imp
29       for _, imp := range r.GetImp() {
30 -         moveRewardedFromPrebidExtTo26(imp)
   +         if err := moveRewardedFromPrebidExtTo26(imp); err != nil {
   +             return err
   +         }
31       }
32   
33       return nil
34   }
35   
36 - // convertUpEnsureExt gets all extension objects required for migration to verify there
37 - // are no access errors.
38 - func convertUpEnsureExt(r *RequestWrapper) error {
39 -     if _, err := r.GetRequestExt(); err != nil {
40 -         return fmt.Errorf("req.ext is invalid: %v", err)
41 -     }
42 - 
43 -     if _, err := r.GetSourceExt(); err != nil {
44 -         return fmt.Errorf("req.source.ext is invalid: %v", err)
45 -     }
46 - 
47 -     if _, err := r.GetRegExt(); err != nil {
48 -         return fmt.Errorf("req.regs.ext is invalid: %v", err)
49 -     }
50 - 
51 -     if _, err := r.GetUserExt(); err != nil {
52 -         return fmt.Errorf("req.user.ext is invalid: %v", err)
53 -     }
54 - 
55 -     for i, imp := range r.GetImp() {
56 -         if _, err := imp.GetImpExt(); err != nil {
57 -             return fmt.Errorf("imp[%v].imp.ext is invalid: %v", i, err)
58 -         }
59 -     }
60 - 
61 -     return nil
62 - }
openrtb_ext/convert_up.go
```

Reply:
> are there plans to use convertUpEnsureExt somewhere else?

I designed the separation of the error checking from the "move "code with the goals of simplifying the main upgrade function and with the hope of being able to eliminate the error checking here entirely since we're downstream of the request validation. Admittedly, this trades off some performance for improved code readability and testability. I felt comfortable with the trade off since I don't believe the inefficiency is significant.

I'm open to a discussion on this though.

> Only traversing the []Imp array once

Yes, I'm taking an inefficiency here and leaning on the fact that the imp array is usually small to help me sleep at night. :)

> Maintain one less function

I actually view this approach as easier to maintain since each move function test no longer needs to deal with the burden of validating the error case.

> Minimize the number of calls to r.GetRequestExt(), r.GetSourceExt(), r.GetRegExt() and r.GetUserExt()

Those do not concern me since the second call to these methods is optimized to return quickly, effectively just returning a cached value.